pageserver/src/tenant/timeline.rs

```diff
@@ -1617,6 +1617,20 @@ impl TraversalLayerExt for Arc<InMemoryLayer> {
     }
 }
 
+enum AnyLayer {
```

Comment:
It still looks like a hack, but it's possible to define an additional `trait` and use it in `dyn`. And it's possible to implement traits in custom types (look at example below).

```rust
use std::sync::Arc;

struct InMemoryLayer { x: String }
struct PersistentLayerImpl { y: String }
trait PersistentLayer {
    fn get_id(&self) -> String;
}

impl PersistentLayer for PersistentLayerImpl {
    fn get_id(&self) -> String {
        self.y.clone()
    }
}

type ArcInMemoryLayer = Arc<InMemoryLayer>;
type ArcDynPersistentLayer = Arc<dyn PersistentLayer>;

trait TraversalIdGetter {
    fn traversal_id(&self) -> String;
}

impl TraversalIdGetter for ArcInMemoryLayer {
    fn traversal_id(&self) -> String {
        self.x.clone()
    }
}

impl TraversalIdGetter for ArcDynPersistentLayer {
    fn traversal_id(&self) -> String {
        self.get_id()
    }
}

fn traversal_id(item: &dyn TraversalIdGetter) -> String {
    item.traversal_id()
}

fn main() {
    let in_memory = Arc::new(InMemoryLayer { x: "in_memory".into() });
    let persistent: ArcDynPersistentLayer = Arc::new(PersistentLayerImpl { y: "persistent".into() });
    
    println!("{}", traversal_id(&in_memory));
    println!("{}", traversal_id(&persistent));
}
```

Reply:
Thank you for suggestion.
Actually there is no need to introduce yet another treat: we already have `TraversalLayerExt`.
So I missed that it is possible to use it instead of `Arc<dyn Layer>`.
Please check new implementation - now AnyLayer enum is removed.

Reply:
much better  👍